mod/wikirate/set/type/source_import_file.rb

```diff
@@ -0,0 +1,228 @@
+    "+year" => { content: "[[#{source_hash[:year]}]]" }
+  )
+end
+
+def create_source source_hash
+  source_card = add_subcard "", type_id: SourceID,
+                                subcards: source_args(source_hash)
+  Env.params[:sourcebox] = "true"
+  source_card.director.catch_up_to_stage :prepare_to_store
+  Env.params[:sourcebox] = nil
+  unless source_card.errors.empty?
+    source_card.errors.each { |k, v| errors.add k, v }
+  end
+  source_card
+end
+
+def create_or_update_subcard source_card, trait, content
+  trait_card = source_card.fetch trait: trait,
+                                 new: { content: "[[#{content}]]" }
+  if trait_card.new?
+    add_subcard trait_card
+  elsif !trait_card.item_names.include?(content)
+    trait_card.add_item content
+    add_subcard trait_card
+  else
+    return false
+  end
+  true
+end
+
+def update_existing_source source_card, source_hash
+  report_type = source_hash[:report_type]
+  updated = create_or_update_subcard source_card, :report_type, report_type
+
+  company = source_hash[:company]
+  updated |= create_or_update_subcard source_card, :wikirate_company, company
+
+  year = source_hash[:year]
+  updated |= create_or_update_subcard source_card, :year, year
+  updated
+end
+
+def update_title_card source_card, source_hash
+  title = Env.params[:title][source_hash[:row].to_s]
+  title_card = source_card.fetch trait: :wikirate_title,
+                                 new: { content: title }
+  if title_card.new?
+    add_subcard title_card
+    return true
+  end
+  false
+end
+
+def handle_duplicated_source source_card, source_hash
+  # title is missing, then add the title
+  #   add company and report type to the related fields
+  #   add to warning message
+  updated = false
+  updated |= update_title_card source_card, source_hash
+  updated |= update_existing_source source_card, source_hash
+  if updated
+    slot_args = success.slot
+    msg_array = [source_hash[:row].to_s, source_card.name]
+    slot_args[:updated_sources].push(msg_array)
+  end
+  nil
+end
+
+def process_source source_hash, source_map
+  url = source_hash[:source]
+  duplicates = Self::Source.find_duplicates url
+  source_card =
+    if duplicates.any?
+      handle_duplicated_source duplicates.first.left, source_hash
+    else
+      create_source source_hash
+    end
+  source_map[url] = source_card
+end
+
+def check_duplication_within_file source_hash, source_map
+  slot_args = success.slot
+  source_url = source_hash[:source]
+  if source_map[source_url]
+    msg = [source_hash[:row].to_s, source_url]
+    slot_args[:duplicated_sources].push(msg)
+    return true
+  end
+  false
+end
+
+def valid_value_data? args
+  @import_errors = []
+  %w(wikirate_company year report_type source).each do |field|
+    add_import_error "#{field} missing", args[:row] if args[field.to_sym].blank?
+  end
+  @import_errors.empty?
+end
+
+# @return updated or created metric value card object
+def parse_source_card source, source_map
```

Comment:
While I am coding, I used to put helper methods above the main method because I can scroll up to the latest written methods easily and continue coding the logic. Nevermind, I will rearrange them.